src/ashell/ide-cmd.c

```diff
@@ -0,0 +1,741 @@
+        ide_spool_flush();
+    }
+    va_start(args, format);
+    size = vsnprintf(spool + spool_cursor, ide_spool_space(), format, args);
+    va_end(args);
+
+    spool_cursor += size;
+    if (spool_cursor >= P_SPOOL_SIZE) {
+        spool_cursor = P_SPOOL_SIZE;
+        ide_spool_flush();
+    }
+    return spool_cursor;
+}
+
+int ide_spool_flush()
+{
+    if (spool_cursor == 0) {
+        return 0;
+    }
+    int size = ide_send_buffer(spool, MIN(spool_cursor, P_SPOOL_SIZE));
+    spool_cursor = 0;
+    // memset(spool, 0, P_SPOOL_SIZE + 1);
+    return size;
+}
+
+int ide_reply(int status, char *message)
+{
+    ide_spool("{\"reply\": \"%s\", \"status\":%d, \"data\": %s }\r\n",
+              cmd_arg_map[parser.cmd_id].cmd, status, message);
+    ide_spool_flush();
+    if (!(status == NO_ERROR && parser.state == PARSE_ARG_STREAM))
+        parser_reset();
+    return -status;
+}
+
+// Functions for building a complete message in the spool before sending.
+int ide_start_message(char *cmd)
+{
+    return ide_spool("{ \"reply\": \"%s\", \"data\": ", cmd);
+}
+
+int ide_end_message(int status)
+{
+    ide_spool(", \"status\": %d }\r\n", status);
+    return ide_spool_flush();
+}
+
+#ifdef ASHELL_IDE_DBG
+#define IDE_DBG(...) do { \
+    ide_spool( __VA_ARGS__ ); \
+    ide_spool_flush(); \
+} while(0)
+#else
+#define IDE_DBG(...)  do { ; } while (0)
+#endif
+
+/* ***************************** Parsing *********************************** */
+
+// Match input to supported commands.
+static ide_cmd_t match_cmd(char *buf, size_t len) {
+    int i = 0;
+    for (; i < CMD_MAX && strncmp(cmd_arg_map[i].cmd, buf, len); i++);
+    return i;  // CMD_MAX means no match
+}
+
+// Consume (tokenize and match) a command from the input.
+static int parse_command(char *buf, size_t len)
+{
+    size_t size, id;
+    char cmd_buf[CMD_MAX_LEN + 1];
+
+    IDE_DBG("\r\nParsing command.");
+    for (size = 0; len > 0 && size <= CMD_MAX_LEN; len--) {
+        if(match_separator(buf) || match_postamble(buf)) {
+            cmd_buf[size] = '\0';
+            id = match_cmd(cmd_buf, size);
+            IDE_DBG("\r\nCommand %d: %s", id, cmd_buf);
+            if (id == CMD_MAX) {
+                return ide_reply(ERROR_INVALID_CMD, "\"Invalid command.\"");
+            }
+            parser.cmd_id = id;
+            return size;
+        }
+        cmd_buf[size++] = *buf++;
+    }
+    return ide_reply(ERROR_INVALID_MSG, "\"Invalid message.\"");
+}
+
+// Check if the number of arguments left to be parsed matches with the input.
+static int check_argc(size_t argc, char *buf, size_t len)
+{
+    IDE_DBG("\r\nChecking %d argument(s).", argc);
+    if (argc == 0 && !match_postamble(buf)) {
+        return ide_reply(ERROR_UNEXPECTED_CHAR, "\"Postamble expected.\"");
+    } else if (argc > 0) {
+        if (match_postamble(buf)) {
+          return ide_reply(ERROR_MISSING_ARG, "\"Missing argument.\"");
+        } else if (!match_separator(buf)) {
```

Comment:
Again I think the postamble check is superfluous - this separator check will also catch it.

Maybe I'm missing something - are these messages to the human IDE user? Is "Missing argument" more helpful to them than "Separator expected"? Do they need to distinguish the cases?